python/lsst/daf/persistence/storageInterface.py

```diff
@@ -0,0 +1,272 @@
+        A list of objects as described by the butler location. One item for
+        each location in butlerLocation.getLocations()
+        """
+
+    @abstractmethod
+    def getLocalFile(self, path):
+        """Get the path to a local copy of the file, downloading it to a
+        temporary if needed.
+
+        Parameters
+        ----------
+        A path the the file in storage, relative to root.
+
+        Returns
+        -------
+        A handle to a local copy of the file. If storage is remote it will be
+        a temporary file. If storage is local it may be the original file or
+        a temporary file. The file name can be gotten via the 'name' property
+        of the returned object.
+        """
+
+    @abstractmethod
+    def exists(self, location):
+        """Check if location exists.
+
+        Parameters
+        ----------
+        location : ButlerLocation or string
+            A a string or a ButlerLocation that describes the location of an
+            object in this storage.
+
+        Returns
+        -------
+        bool
+            True if exists, else False.
+        """
+
+    @abstractmethod
+    def instanceSearch(self, path):
+        """Search for the given path in this storage instance.
+
+        If the path contains an HDU indicator (a number in brackets before the
+        dot, e.g. 'foo.fits[1]', this will be stripped when searching and so
+        will match filenames without the HDU indicator, e.g. 'foo.fits'. The
+        path returned WILL contain the indicator though, e.g. ['foo.fits[1]'].
+
+        Parameters
+        ----------
+        path : string
+            A filename (and optionally prefix path) to search for within root.
+
+        Returns
+        -------
+        string or None
+            The location that was found, or None if no location was found.
+        """
+
+    @staticmethod
+    @abstractmethod
+    def search(root, path):
+        """Look for the given path in the current root.
+
+        Also supports searching for the path in Butler v1 repositories by
+        following the Butler v1 _parent symlink
+
+        If the path contains an HDU indicator (a number in brackets, e.g.
+        'foo.fits[1]', this will be stripped when searching and so
+        will match filenames without the HDU indicator, e.g. 'foo.fits'. The
+        path returned WILL contain the indicator though, e.g. ['foo.fits[1]'].
+
+        Parameters
+        ----------
+        root : string
+            The path to the root directory.
+        path : string
+            The path to the file within the root directory.
+
+        Returns
+        -------
+        string or None
+            The location that was found, or None if no location was found.
+        """
+
+    @abstractmethod
+    def copyFile(self, fromLocation, toLocation):
+        """Copy a file from one location to another on the local filesystem.
+
+        Parameters
+        ----------
+        fromLocation : path
```

Comment:
is `path` a `str`? I think we are supposed to list actual Python types on the right side.